## Khan Academy Math Student Dashboard: Data Cleaning
We will be creating a student dashboard that displays each student's progress through Khan Academy Math. The goal is to showcase:
- How the student is performing overall and this month?
- What is their assignment completion rate?
- Which topics are students strongest and weakest in?
- The students' percentile in the class?
- Is the student growing?

We want to automate the dashboarding process for each class in Lyceum Village, grades K-8. We will be testing on grade K-1 students' Khan Academy Math first to create functions and a webpage to showcase their progress.

### Import the Data
The data imported is for grade K-1. The students' names are already adjusted to an alias name.

In [1]:
import numpy as np
import pandas as pd
import datetime
import re

In [2]:
df = pd.read_csv("../Resources/K_1-Math-Anonymous.csv")

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Assignment Name              266 non-null    object 
 1   Student Name                 266 non-null    object 
 2   Score At Due Date            26 non-null     float64
 3   Score Best Ever              30 non-null     float64
 4   Points Possible              30 non-null     float64
 5   Number Of Attempts           266 non-null    int64  
 6   Most Recent Completion Date  30 non-null     object 
 7   Start Date                   266 non-null    object 
 8   Due Date                     266 non-null    object 
 9   Assignment URL               266 non-null    object 
 10  Assignment Type              266 non-null    object 
 11  Due Date (no time)           266 non-null    object 
dtypes: float64(3), int64(1), object(8)
memory usage: 25.1+ KB


Our csv dataset will have 11 variables.

Types of data:
- 3 float: 
    - Score At Due Date
    - Score Best Ever
    - Points Possible
- 1 integer:
    - Number Of Attempts
- 7 string:
    - Assignment Name
    - Student Name
    - Most Recent Completion Date
    - Start Date
    - Due Date
    - Assignment URL
    - Assignment Type

The data type so far matches what they should be.

There are 266 rows total. There are missing data in the following columns:
- Score At Due Date: 240 missing
- Score Best Ever: 224 missing
- Points Possible: 224 missing
- Most Recent Completion Date: 224 missing

They are missing data because some students did not complete their assignments. We will be removing these data in actual student analysis.

### Student Anonymity
We will need to create a student ID column to give each student anonymity on their scores when it is being displayed to other parents.
Their ID's will be formatted as so: grade_studentinitial_number <br>

Example: G2_KL_1

#### Create Student ID
We will be automating this system by having it done automatically only once. The application will request the grade level from the user and store the student ID information into a PostgreSQL database.

In [4]:
df['Student Name'].unique()

array(['Ned Ethans', 'Anna Kite', 'April Luna', 'Andy Hong',
       'Abby Nguyen', 'Everett Chase', 'Kyle Anderson', 'Kevin Martelle',
       'Misa Bing', 'Olsen Le', 'Terry Long'], dtype=object)

In [5]:
# User input grade level
stu_grade = input("Enter the student grade level: ")

In [6]:
# Loop through the list of unique students and create their ID
stu_id_list = []
for name in df['Student Name']:
    stu_name = name.split(" ")
    stu_id = "G" + stu_grade + "_" + stu_name[0][0] + stu_name[1][0]
    stu_id_list.append(stu_id)
stu_id_list[:5]

['GK1_NE', 'GK1_AK', 'GK1_AL', 'GK1_AH', 'GK1_AN']

In [7]:
# Add in the numeric value to account for students with the same initials
stu_id_list2 = []
if len(df['Student Name']) == len(stu_id_list):
    for id in stu_id_list:
        # Add 1 at the end because there are no repeated initials
        stu_id_list2.append(id + "1")
else:
    # change the number depending on the student name
    # current list of students do not have repeated initials. Create a dummy list to test this function on
    stu_id_list2
stu_id_list2[:5]

['GK1_NE1', 'GK1_AK1', 'GK1_AL1', 'GK1_AH1', 'GK1_AN1']

In [8]:
# Add id column to the front of the dataframe
df.insert(loc=0, column='Student ID', value=stu_id_list2)
df.head()

Student ID          Assignment Name Student Name  Score At Due Date  \
0    GK1_NE1  Comparing numbers to 10   Ned Ethans                NaN   
1    GK1_AK1            Add within 10    Anna Kite                NaN   
2    GK1_AL1            Add within 10   April Luna                NaN   
3    GK1_AH1            Add within 10    Andy Hong                NaN   
4    GK1_AN1            Add within 10  Abby Nguyen                NaN   

   Score Best Ever  Points Possible  Number Of Attempts  \
0              NaN              NaN                   0   
1              NaN              NaN                   0   
2              NaN              NaN                   0   
3              NaN              NaN                   0   
4              NaN              NaN                   0   

  Most Recent Completion Date        Start Date          Due Date  \
0                         NaN  Sep 29th, 4:50PM  Oct 8th, 11:59PM   
1                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
2                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
3                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
4                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   

                                      Assignment URL Assignment Type  \
0  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
1  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
2  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
3  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
4  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

  Due Date (no time)  
0          Oct 8th,   
1          Oct 8th,   
2          Oct 8th,   
3          Oct 8th,   
4          Oct 8th,

## Missing Data
NaN in the Score At Due Date, Score Best Ever, and Points Possible columns means that students did not complete their assignment. Thus, we will need to create a column that states the completion of their assignment.

Create a column with these values:
- Missing: Has 0 as the Number Of Attempts
- Complete: Has 1 or higher Number of Attempts and the Most Recent Completion Date is before or on the Due Date
- Late: Has 1 or more Number Of Attempts and the Most Recent Completion Date is past the Due Date

### Convert Date Columns to Date type
Convert the two columns from string object to date:
- Most Recent Completion Date
- Due Date

In [9]:
# Find the rows with no missing data
df_val = df[df['Most Recent Completion Date'].isna()==False]
df_val.head()

Student ID      Assignment Name    Student Name  Score At Due Date  \
96     GK1_KM1  Counting: Unit test  Kevin Martelle               13.0   
100    GK1_TL1  Counting: Unit test      Terry Long               13.0   
107    GK1_KM1         Add within 5  Kevin Martelle                7.0   
111    GK1_TL1         Add within 5      Terry Long                7.0   
118    GK1_KM1    Subtract within 5  Kevin Martelle                7.0   

     Score Best Ever  Points Possible  Number Of Attempts  \
96              13.0             13.0                   2   
100             13.0             13.0                   1   
107              7.0              7.0                   1   
111              7.0              7.0                   1   
118              7.0              7.0                   1   

    Most Recent Completion Date         Start Date           Due Date  \
96             Sep 20th, 4:55PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
100            Sep 20th, 4:42PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
107            Sep 20th, 5:01PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
111            Sep 20th, 4:50PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
118            Sep 20th, 5:04PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   

                                        Assignment URL Assignment Type  \
96   https://www.khanacademy.org/math/early-math/cc...       Unit Test   
100  https://www.khanacademy.org/math/early-math/cc...       Unit Test   
107  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
111  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
118  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

    Due Date (no time)  
96          Sep 24th,   
100         Sep 24th,   
107         Sep 24th,   
111         Sep 24th,   
118         Sep 24th,

In [10]:
# Function to convert the date from string type to date type
def convertDate(date_str, year):
    split_date = re.split(", | ", date_str) # Splits the date into Month, Day, Time
    day_num = int(re.findall('\d+', split_date[1])[0]) # Removes letters from the day. Example: 8th becomes 8
    month_num = datetime.datetime.strptime(split_date[0], '%b').month # Converts the month into the month's number
    
    # we are missing the year; we will have to have the user input it
    # Combined date
    date_str = str(month_num) + "/" + str(day_num) + "/" + year + " " + split_date[2]
    comb_date = datetime.datetime.strptime(date_str, '%m/%d/%Y %H:%M%p')

    return comb_date
    

In [11]:
progress_list = []
index_attempts = df.columns.get_loc('Number Of Attempts')
index_duedate = df.columns.get_loc('Due Date')
index_compdate = df.columns.get_loc('Most Recent Completion Date')
report_year = input('Year of report: ')
for i, r in df.iterrows():
    if r[index_attempts] == 0:
        progress_list.append('Missing')
    else:
        duedate_formatted = convertDate(r[index_duedate], report_year)
        compdate_formatted = convertDate(r[index_compdate], report_year)
        if duedate_formatted > compdate_formatted:
            progress_list.append('Complete')
        else:
            progress_list.append('Late')

In [12]:
# Add Progress column into the df
df['Progress'] = progress_list
df.head()

Student ID          Assignment Name Student Name  Score At Due Date  \
0    GK1_NE1  Comparing numbers to 10   Ned Ethans                NaN   
1    GK1_AK1            Add within 10    Anna Kite                NaN   
2    GK1_AL1            Add within 10   April Luna                NaN   
3    GK1_AH1            Add within 10    Andy Hong                NaN   
4    GK1_AN1            Add within 10  Abby Nguyen                NaN   

   Score Best Ever  Points Possible  Number Of Attempts  \
0              NaN              NaN                   0   
1              NaN              NaN                   0   
2              NaN              NaN                   0   
3              NaN              NaN                   0   
4              NaN              NaN                   0   

  Most Recent Completion Date        Start Date          Due Date  \
0                         NaN  Sep 29th, 4:50PM  Oct 8th, 11:59PM   
1                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
2                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
3                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
4                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   

                                      Assignment URL Assignment Type  \
0  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
1  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
2  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
3  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
4  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

  Due Date (no time) Progress  
0          Oct 8th,   Missing  
1          Oct 8th,   Missing  
2          Oct 8th,   Missing  
3          Oct 8th,   Missing  
4          Oct 8th,   Missing

The Progress column which indicates if the assignment is missing, late, or complete. Now we can convert all missing data from the following columns into 0 if they are marked "Missing":
- Score At Due Date
- Score Best Ever

In [13]:
df.loc[df['Progress'] == 'Missing', ['Score At Due Date', 'Score Best Ever']] = 0
df.head()

Student ID          Assignment Name Student Name  Score At Due Date  \
0    GK1_NE1  Comparing numbers to 10   Ned Ethans                0.0   
1    GK1_AK1            Add within 10    Anna Kite                0.0   
2    GK1_AL1            Add within 10   April Luna                0.0   
3    GK1_AH1            Add within 10    Andy Hong                0.0   
4    GK1_AN1            Add within 10  Abby Nguyen                0.0   

   Score Best Ever  Points Possible  Number Of Attempts  \
0              0.0              NaN                   0   
1              0.0              NaN                   0   
2              0.0              NaN                   0   
3              0.0              NaN                   0   
4              0.0              NaN                   0   

  Most Recent Completion Date        Start Date          Due Date  \
0                         NaN  Sep 29th, 4:50PM  Oct 8th, 11:59PM   
1                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
2                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
3                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
4                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   

                                      Assignment URL Assignment Type  \
0  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
1  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
2  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
3  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
4  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

  Due Date (no time) Progress  
0          Oct 8th,   Missing  
1          Oct 8th,   Missing  
2          Oct 8th,   Missing  
3          Oct 8th,   Missing  
4          Oct 8th,   Missing

Score At Due Date is zero for late assignments. Need to change all NaN to 0 if it is late.

In [14]:
df.loc[df['Progress'] == 'Late', ['Score At Due Date']] = 0

The following columns contains missing data for missed assignments, but cannot be set to 0:
- Points Possible: this is the total points received for the assignment. It should be set depending on the points available per assignment.
- Most Recent Completion Date: this will be left as missing as it is unrelated to scores

We need to find the Points Possible value for each assignment, and fill in the missing information.

In [15]:
df_no_missing = df.loc[df['Points Possible'].isna() == False]
df_no_missing.head()

Student ID      Assignment Name    Student Name  Score At Due Date  \
96     GK1_KM1  Counting: Unit test  Kevin Martelle               13.0   
100    GK1_TL1  Counting: Unit test      Terry Long               13.0   
107    GK1_KM1         Add within 5  Kevin Martelle                7.0   
111    GK1_TL1         Add within 5      Terry Long                7.0   
118    GK1_KM1    Subtract within 5  Kevin Martelle                7.0   

     Score Best Ever  Points Possible  Number Of Attempts  \
96              13.0             13.0                   2   
100             13.0             13.0                   1   
107              7.0              7.0                   1   
111              7.0              7.0                   1   
118              7.0              7.0                   1   

    Most Recent Completion Date         Start Date           Due Date  \
96             Sep 20th, 4:55PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
100            Sep 20th, 4:42PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
107            Sep 20th, 5:01PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
111            Sep 20th, 4:50PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
118            Sep 20th, 5:04PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   

                                        Assignment URL Assignment Type  \
96   https://www.khanacademy.org/math/early-math/cc...       Unit Test   
100  https://www.khanacademy.org/math/early-math/cc...       Unit Test   
107  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
111  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
118  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

    Due Date (no time)  Progress  
96          Sep 24th,   Complete  
100         Sep 24th,   Complete  
107         Sep 24th,   Complete  
111         Sep 24th,   Complete  
118         Sep 24th,   Complete

In [16]:
# Check if the assignments in the dataframe with no missing values is the same for those that are missing.
df_nm_uniq = df.loc[df['Points Possible'].isna() == False, ['Assignment Name']]['Assignment Name'].unique()
df_nm_uniq

array(['Counting: Unit test', 'Add within 5', 'Subtract within 5',
       'Count objects 2', 'Compare numbers of objects 1',
       'Comparing numbers to 10', 'Counting: Quiz 2', 'Numbers to 100',
       'Numbers to 120', 'Count tens', 'Counting: Quiz 1',
       'Count with small numbers', 'Count in order',
       'Find 1 more or 1 less than a number', 'Missing numbers'],
      dtype=object)

In [17]:
# List of unique Assignment Names from values with missing data
df_m_uniq = df.loc[df['Points Possible'].isna() == True, ['Assignment Name']]['Assignment Name'].unique()
df_m_uniq

array(['Comparing numbers to 10', 'Add within 10', 'Subtract within 10',
       'Addition and subtraction intro: Quiz 2',
       'Relate addition and subtraction', 'Making 5',
       'Making small numbers in different ways',
       'Addition and subtraction intro: Quiz 1',
       'Make 10 (grids and number bonds)', 'Counting: Unit test',
       'Add within 5', 'Subtract within 5', 'Count objects 2',
       'Compare numbers of objects 1', 'Compare numbers of objects 2',
       'Counting: Quiz 2', 'Numbers to 100', 'Numbers to 120',
       'Count tens', 'Counting: Quiz 1', 'Count with small numbers',
       'Count in order', 'Find 1 more or 1 less than a number',
       'Missing numbers'], dtype=object)

In [18]:
print(f"Total Assignments that students did not complete: {len(df_m_uniq)}")
print(f"Total Assignments that students did complete: {len(df_nm_uniq)}")

Total Assignments that students did not complete: 24
Total Assignments that students did complete: 15


In [19]:
# Print out the assignment names that are missing and not in the completed assignment list
all_missing_assignments = []
for assignment in df_m_uniq:
    if assignment not in df_nm_uniq:
        all_missing_assignments.append(assignment)
all_missing_assignments

['Add within 10',
 'Subtract within 10',
 'Addition and subtraction intro: Quiz 2',
 'Relate addition and subtraction',
 'Making 5',
 'Making small numbers in different ways',
 'Addition and subtraction intro: Quiz 1',
 'Make 10 (grids and number bonds)',
 'Compare numbers of objects 2']

The assignments listed above are missing for all students in the class. Therefore, we would have to manually input the Points Possible. Thus, we will just fill in these assignments with 10 possible points.

In [20]:
# Fill in the Points Possible value as 10 for assignments that all students are missing
df.loc[df['Assignment Name'].isin(all_missing_assignments) == True, 'Points Possible'] = 10
df.head(20)

Student ID          Assignment Name    Student Name  Score At Due Date  \
0     GK1_NE1  Comparing numbers to 10      Ned Ethans                0.0   
1     GK1_AK1            Add within 10       Anna Kite                0.0   
2     GK1_AL1            Add within 10      April Luna                0.0   
3     GK1_AH1            Add within 10       Andy Hong                0.0   
4     GK1_AN1            Add within 10     Abby Nguyen                0.0   
5     GK1_EC1            Add within 10   Everett Chase                0.0   
6     GK1_KA1            Add within 10   Kyle Anderson                0.0   
7     GK1_KM1            Add within 10  Kevin Martelle                0.0   
8     GK1_MB1            Add within 10       Misa Bing                0.0   
9     GK1_NE1            Add within 10      Ned Ethans                0.0   
10    GK1_OL1            Add within 10        Olsen Le                0.0   
11    GK1_TL1            Add within 10      Terry Long                0.0   
12    GK1_AK1       Subtract within 10       Anna Kite                0.0   
13    GK1_AL1       Subtract within 10      April Luna                0.0   
14    GK1_AH1       Subtract within 10       Andy Hong                0.0   
15    GK1_AN1       Subtract within 10     Abby Nguyen                0.0   
16    GK1_EC1       Subtract within 10   Everett Chase                0.0   
17    GK1_KA1       Subtract within 10   Kyle Anderson                0.0   
18    GK1_KM1       Subtract within 10  Kevin Martelle                0.0   
19    GK1_MB1       Subtract within 10       Misa Bing                0.0   

    Score Best Ever  Points Possible  Number Of Attempts  \
0               0.0              NaN                   0   
1               0.0             10.0                   0   
2               0.0             10.0                   0   
3               0.0             10.0                   0   
4               0.0             10.0                   0   
5               0.0             10.0                   0   
6               0.0             10.0                   0   
7               0.0             10.0                   0   
8               0.0             10.0                   0   
9               0.0             10.0                   0   
10              0.0             10.0                   0   
11              0.0             10.0                   0   
12              0.0             10.0                   0   
13              0.0             10.0                   0   
14              0.0             10.0                   0   
15              0.0             10.0                   0   
16              0.0             10.0                   0   
17              0.0             10.0                   0   
18              0.0             10.0                   0   
19              0.0             10.0                   0   

   Most Recent Completion Date        Start Date          Due Date  \
0                          NaN  Sep 29th, 4:50PM  Oct 8th, 11:59PM   
1                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
2                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
3                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
4                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
5                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
6                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
7                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
8                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
9                          NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
10                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
11                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
12                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
13                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
14                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
15  

In [21]:
# Find the Points Possible for missing assignments where some students have completed it.
df_names_points = df_no_missing.loc[df_no_missing['Assignment Name'].isin(all_missing_assignments)==False, ['Assignment Name', 'Points Possible']].drop_duplicates()
df_names_points

Assignment Name  Points Possible
96                   Counting: Unit test             13.0
107                         Add within 5              7.0
118                    Subtract within 5              7.0
131                      Count objects 2              7.0
142         Compare numbers of objects 1              7.0
153              Comparing numbers to 10              7.0
173                     Counting: Quiz 2              6.0
178                       Numbers to 100              7.0
192                       Numbers to 120              7.0
206                           Count tens              7.0
217                     Counting: Quiz 1              7.0
222             Count with small numbers              7.0
233                       Count in order              7.0
250  Find 1 more or 1 less than a number              7.0
261                      Missing numbers              7.0

In [22]:
for index, row in df.iterrows():
    if row['Assignment Name'] in df_nm_uniq:
        print(df['Points Possible'][index])
        print(df_names_points.loc[df_names_points['Assignment Name'] == df['Assignment Name'][index], 'Points Possible'])
    break

nan
153    7.0
Name: Points Possible, dtype: float64


In [23]:
# Fill in missing values with the correct corresponding Points Possible
for index, row in df.iterrows():
    if row['Assignment Name'] in df_nm_uniq:
        df['Points Possible'][index] = df_names_points.loc[df_names_points['Assignment Name'] == df['Assignment Name'][index], 'Points Possible']
df.loc[df['Assignment Name'].isin(df_nm_uniq) == True, ['Assignment Name', 'Points Possible']]

C:\Users\krist\AppData\Local\Temp\ipykernel_21020\1929856247.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Points Possible'][index] = df_names_points.loc[df_names_points['Assignment Name'] == df['Assignment Name'][index], 'Points Possible']


Assignment Name  Points Possible
0    Comparing numbers to 10              7.0
89   Comparing numbers to 10              7.0
90       Counting: Unit test             13.0
91       Counting: Unit test             13.0
92       Counting: Unit test             13.0
..                       ...              ...
261          Missing numbers              7.0
262          Missing numbers              7.0
263          Missing numbers              7.0
264          Missing numbers              7.0
265          Missing numbers              7.0

[167 rows x 2 columns]

In [24]:
# Check rows to see if the missing numbers were filled properly for Possible Points
df.loc[df['Assignment Name'].isin(df_nm_uniq)==True, ['Assignment Name', 'Points Possible']].drop_duplicates()

Assignment Name  Points Possible
0                Comparing numbers to 10              7.0
90                   Counting: Unit test             13.0
101                         Add within 5              7.0
112                    Subtract within 5              7.0
123                      Count objects 2              7.0
134         Compare numbers of objects 1              7.0
167                     Counting: Quiz 2              6.0
178                       Numbers to 100              7.0
189                       Numbers to 120              7.0
200                           Count tens              7.0
211                     Counting: Quiz 1              7.0
222             Count with small numbers              7.0
233                       Count in order              7.0
244  Find 1 more or 1 less than a number              7.0
255                      Missing numbers              7.0

In [25]:
# Check if there are any missing values in Possible Points
df['Points Possible'].isna().sum()

0

## Add Assignment Grade Level Column
Although the class is filled with kindergarten and 1st graders, the assignments vary in difficulty. We will need to find the unique assignment levels from the URL.

In [26]:
# Allows us to print entire value without truncating
pd.set_option('display.max_colwidth', -1)
print(df["Assignment URL"].head(5))

0    https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-counting/kindergarten-comparing-numbers/e/comparing-numbers-through-10
1    https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together               
2    https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together               
3    https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together               
4    https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together               
Name: Assignment URL, dtype: object


C:\Users\krist\AppData\Local\Temp\ipykernel_21020\1367671070.py:2: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth', -1)


The unit topics are located after "cc-kindergarten-". We can extract to collect the unit topic by splitting the URL using "cc-".

In [27]:
unit_topic = []
for rows in df["Assignment URL"]:
    unit_topic.append(re.split("/cc-|#cc-|/", rows)[5])

In [28]:
uniq_unit_topic = set(unit_topic)
print(uniq_unit_topic)

{'1st-add-subtract', 'kindergarten-add-subtract', 'early-math-counting-topic', 'early-math-add-sub-basics', 'kindergarten-counting', '1st-place-value'}


The different levels of difficulty are displayed as the first word in the string, and then the unit topic follows after. We will now split this string and add them to the dataframe as two new columns called "Assignment Grade Level" and "Unit Topic".

In [29]:
level_list = []
unit_list = []

for rows in unit_topic:
    if("early-math" in rows):
        pattern = r'^([\w]+-[\w]+)'
        row_split = re.match(pattern, rows).group()
        level_list.append(row_split)
        
        if ("add-sub" in rows[len(row_split)+1:]):
            unit_list.append("add-subtract")
        elif ("counting" in rows[len(row_split)+1:]):
            unit_list.append("counting")
    else:
        row_split = rows.split("-", 1)
        level_list.append(row_split[0])
        unit_list.append(row_split[1])

print(len(level_list), len(unit_list))
level_set = set(level_list)
unit_set = set(unit_list)
print(level_set)
print(unit_set)

266 266
{'kindergarten', 'early-math', '1st'}
{'add-subtract', 'place-value', 'counting'}


In [30]:
# Add Grade Level and Unit Topic to dataframe
df['Assignment Grade Level'] = level_list
df['Unit Topic'] = unit_list
df.head()

Student ID          Assignment Name Student Name  Score At Due Date  \
0  GK1_NE1    Comparing numbers to 10  Ned Ethans   0.0                 
1  GK1_AK1    Add within 10            Anna Kite    0.0                 
2  GK1_AL1    Add within 10            April Luna   0.0                 
3  GK1_AH1    Add within 10            Andy Hong    0.0                 
4  GK1_AN1    Add within 10            Abby Nguyen  0.0                 

   Score Best Ever  Points Possible  Number Of Attempts  \
0  0.0              7.0              0                    
1  0.0              10.0             0                    
2  0.0              10.0             0                    
3  0.0              10.0             0                    
4  0.0              10.0             0                    

  Most Recent Completion Date        Start Date          Due Date  \
0  NaN                         Sep 29th, 4:50PM  Oct 8th, 11:59PM   
1  NaN                         Sep 29th, 4:48PM  Oct 8th, 11:59PM   
2  NaN                         Sep 29th, 4:48PM  Oct 8th, 11:59PM   
3  NaN                         Sep 29th, 4:48PM  Oct 8th, 11:59PM   
4  NaN                         Sep 29th, 4:48PM  Oct 8th, 11:59PM   

                                                                                                                                 Assignment URL  \
0  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-counting/kindergarten-comparing-numbers/e/comparing-numbers-through-10   
1  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together                  
2  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together                  
3  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together                  
4  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/put-together                  

  Assignment Type Due Date (no time) Progress Assignment Grade Level  \
0  Exercise        Oct 8th,           Missing  kindergarten            
1  Exercise        Oct 8th,           Missing  kindergarten            
2  Exercise        Oct 8th,           Missing  kindergarten            
3  Exercise        Oct 8th,           Missing  kindergarten            
4  Exercise        Oct 8th,           Missing  kindergarten            

     Unit Topic  
0  counting      
1  add-subtract  
2  add-subtract  
3  add-subtract  
4  add-subtract

## Weighted Scores
Khan Academy organizes student progress based on completion. However, this doesn't tell much about students' math comprehension. Therefore, we will be creating a measure to more accurately gauge student comprehension. Therefore, each score will be weighted by the number of attempts for each assignment.

Formula:

$$
Weighted Score = Actual Score (1 + \frac{Class Avg Attempts - Student Total Attempts}{Class Avg Attempts})
$$ (my_other_label)

Where ClassAvgAttempts is the average number of attempts for the assignment.

In [31]:
# Find the ClassAvgAttempts per assignment
# Ignore the rows where students did not complete the assignment
df_mean = df.loc[df['Progress'] != 'Missing'].groupby(['Assignment Name']).mean()
df_mean

Score At Due Date  Score Best Ever  \
Assignment Name                                                           
Add within 5                         7.000000           7.000000          
Compare numbers of objects 1         6.500000           6.500000          
Comparing numbers to 10              4.000000           4.000000          
Count in order                       4.666667           7.000000          
Count objects 2                      7.000000           7.000000          
Count tens                           7.000000           7.000000          
Count with small numbers             4.666667           6.666667          
Counting: Quiz 1                     0.000000           7.000000          
Counting: Quiz 2                     0.000000           6.000000          
Counting: Unit test                  13.000000          13.000000         
Find 1 more or 1 less than a number  6.000000           6.000000          
Missing numbers                      6.500000           6.500000          
Numbers to 100                       6.666667           6.666667          
Numbers to 120                       3.000000           6.500000          
Subtract within 5                    7.000000           7.000000          

                                     Points Possible  Number Of Attempts  
Assignment Name                                                           
Add within 5                         7.0              1.000000            
Compare numbers of objects 1         7.0              2.500000            
Comparing numbers to 10              7.0              1.000000            
Count in order                       7.0              2.000000            
Count objects 2                      7.0              1.000000            
Count tens                           7.0              1.000000            
Count with small numbers             7.0              1.000000            
Counting: Quiz 1                     7.0              2.000000            
Counting: Quiz 2                     6.0              1.000000            
Counting: Unit test                  13.0             1.500000            
Find 1 more or 1 less than a number  7.0              3.000000            
Missing numbers                      7.0              2.000000            
Numbers to 100                       7.0              1.333333            
Numbers to 120                       7.0              1.500000            
Subtract within 5                    7.0              1.000000

Let's create a function that standardizes scores based on the averages. We will call this function `weight_Score_attempts`. This function will return a new dataframe with the score weighted by the average of attempts.
<br>`weight_Score_attempts(df, mean_df)`
<br>Arguments:
- `df` (dataframe): dataframe that includes the student scores.
- `mean_df` (dataframe): dataframe that includes the mean values of all numerical columns


In [32]:
def weight_Score_attempts(df, mean_df):
    df_new = df.merge(mean_df[['Number Of Attempts']], on= "Assignment Name", how='outer')    
    df_new['Weighted Score'] = (1 + (df_new['Number Of Attempts_y'] - df_new['Number Of Attempts_x'])/df_new['Number Of Attempts_y'])*df_new["Score Best Ever"]
    # Have any missing data for Number Of Attempts_y and Weighted Score set to 0
    df_new.fillna({'Number Of Attempts_y':0, 'Weighted Score': 0}, inplace=True)
    df_new.columns= list(df.columns) + ['Average Number Of Attempts', 'Weighted Score']
    return df_new

In [33]:
df_weight = weight_Score_attempts(df, df_mean)

In [34]:
len(df_weight.loc[df_weight['Weighted Score'].isna() == True,])

0

In [35]:
df_weight.sample(10)

Student ID                         Assignment Name    Student Name  \
145  GK1_AK1    Compare numbers of objects 1            Anna Kite        
210  GK1_TL1    Count tens                              Terry Long       
118  GK1_KM1    Add within 5                            Kevin Martelle   
85   GK1_KM1    Addition and subtraction intro: Quiz 1  Kevin Martelle   
41   GK1_KM1    Addition and subtraction intro: Quiz 2  Kevin Martelle   
32   GK1_NE1    Subtract within 10                      Ned Ethans       
131  GK1_NE1    Subtract within 5                       Ned Ethans       
68   GK1_AK1    Making small numbers in different ways  Anna Kite        
251  GK1_MB1    Find 1 more or 1 less than a number     Misa Bing        
178  GK1_AK1    Numbers to 100                          Anna Kite        

     Score At Due Date  Score Best Ever  Points Possible  Number Of Attempts  \
145  0.0                0.0              7.0              0                    
210  0.0                0.0              7.0              0                    
118  7.0                7.0              7.0              1                    
85   0.0                0.0              10.0             0                    
41   0.0                0.0              10.0             0                    
32   0.0                0.0              10.0             0                    
131  0.0                0.0              7.0              0                    
68   0.0                0.0              10.0             0                    
251  0.0                0.0              7.0              0                    
178  6.0                6.0              7.0              1                    

    Most Recent Completion Date         Start Date           Due Date  \
145  NaN                         Sep 3rd, 1:47PM    Sep 17th, 11:59PM   
210  NaN                         Sep 1st, 1:04PM    Sep 10th, 11:59PM   
118  Sep 20th, 5:01PM            Sep 10th, 10:33AM  Sep 24th, 11:59PM   
85   NaN                         Sep 29th, 4:48PM   Oct 1st, 11:59PM    
41   NaN                         Sep 29th, 4:48PM   Oct 8th, 11:59PM    
32   NaN                         Sep 29th, 4:48PM   Oct 8th, 11:59PM    
131  NaN                         Sep 10th, 10:33AM  Sep 24th, 11:59PM   
68   NaN                         Sep 29th, 4:48PM   Oct 1st, 11:59PM    
251  NaN                         Aug 25th, 9:35AM   Sep 3rd, 11:59PM    
178  Sep 7th, 5:18PM             Sep 1st, 1:04PM    Sep 10th, 11:59PM   

                                                                                                                                          Assignment URL  \
145  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-counting/kindergarten-comparing-numbers/e/compare-groups-through-10             
210  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-counting/kindergarten-numbers-100/e/counting-tens                               
118  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-add-sub-intro/e/addition_1                            
85   https://www.khanacademy.org/math/early-math/cc-early-math-add-sub-basics#cc-early-math-add-sub-basics-quiz-1                                          
41   https://www.khanacademy.org/math/early-math/cc-early-math-add-sub-basics#cc-early-math-add-sub-basics-quiz-2                                          
32   https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-together-apart/e/take-apart                           
131  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-add-sub-intro/e/subtraction_1                         
68   https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-add-subtract/kindergarten-make-10/e/making-totals-in-different-ways-within-10   
251  https://www.khanacademy.org/math/cc-kindergarten-math/cc-kindergarten-counting/kindergarten-cou

## Remove Unnecessary Columns
Now that our dataset it clean, let's remove any unnecessary columns before importing it into a SQL database. The columns that will need to be removed are:
- Score At Due Date: We can remove this as we will be focusing on their Best Score Ever
- Assignment URL: We have extracted the necessary information from the URL already so we no longer need this section
- Due Date (no time): We already have a Due Date column so this section is now unnecessary

In [36]:
df_clean = df.drop(columns=['Score At Due Date', 'Assignment URL', 'Due Date (no time)'], axis=1)
df_clean.columns

Index(['Student ID', 'Assignment Name', 'Student Name', 'Score Best Ever',
       'Points Possible', 'Number Of Attempts', 'Most Recent Completion Date',
       'Start Date', 'Due Date', 'Assignment Type', 'Progress',
       'Assignment Grade Level', 'Unit Topic'],
      dtype='object')

## PostgreSQL
We will now be saving our data within a Postgresql database called `student_dashboard_db`. Our database already has two precreated data tables:
1. `student_info`: Contains all student names, ID's, and grade levels
2. `student_progress`: Contains all of the student's progress in Khan Academy Math

The goal is to be able to:
- Generate unique student ID's for new students and maintaining student ID's for those who already have one
- Store cleaned data for student progress that we can use to generate an interactive Student Dashboard

Let's connect Python to our student_dashboard_db.


In [62]:
# Import psycopg2 which allows Python to connect and make changes to PostrgreSQL
import psycopg2
from sqlalchemy import create_engine
from db_info import db, db_user, db_host, db_password, db_port

In [63]:
# Establish connection
conn = psycopg2.connect(
    database=db, user = db_user, password = db_password, host = db_host, port = db_port
)
cursor = conn.cursor()

In [69]:
db_engine = 'postgresql+psycopg2://' + db_user +':'+ db_password + '@' + db_host + ':'+ db_port +'/'+ db
engine = create_engine(db_engine)
conn_db = engine.connect()

In [70]:
conn.autocommit = True
cursor = conn.cursor()

Our next step is to update the student_info and student_progress table with the Khan Academy Math data we cleaned.

### Updating student_info Table
Our student data table should contain student information to help identify students when creating our dashboard. They will include the following columns:
- Name
- Unique Student ID
- Grade Level

In [39]:
# Create a new dataframe with only the specified information
stu_info_df = df_clean[['Student Name', 'Student ID']].drop_duplicates()
# Add Student Grade Level Column
stu_info_df['Student Grade Level'] = "K-1"
stu_info_df

Student Name Student ID Student Grade Level
0   Ned Ethans      GK1_NE1    K-1               
1   Anna Kite       GK1_AK1    K-1               
2   April Luna      GK1_AL1    K-1               
3   Andy Hong       GK1_AH1    K-1               
4   Abby Nguyen     GK1_AN1    K-1               
5   Everett Chase   GK1_EC1    K-1               
6   Kyle Anderson   GK1_KA1    K-1               
7   Kevin Martelle  GK1_KM1    K-1               
8   Misa Bing       GK1_MB1    K-1               
10  Olsen Le        GK1_OL1    K-1               
11  Terry Long      GK1_TL1    K-1

In [71]:
# Convert the dataframe to SQL where student_info is the data table
stu_info_df.to_sql(name = 'student_info', con=engine, if_exists='append', index=False)

ProgrammingError: (psycopg2.errors.UndefinedColumn) column "Student Name" of relation "student_info" does not exist
LINE 1: INSERT INTO student_info ("Student Name", "Student ID", "Stu...
                                  ^

[SQL: INSERT INTO student_info ("Student Name", "Student ID", "Student Grade Level") VALUES (%(Student Name)s, %(Student ID)s, %(Student Grade Level)s)]
[parameters: ({'Student Name': 'Ned Ethans', 'Student ID': 'GK1_NE1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Anna Kite', 'Student ID': 'GK1_AK1', 'Student Grade Level': 'K-1'}, {'Student Name': 'April Luna', 'Student ID': 'GK1_AL1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Andy Hong', 'Student ID': 'GK1_AH1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Abby Nguyen', 'Student ID': 'GK1_AN1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Everett Chase', 'Student ID': 'GK1_EC1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Kyle Anderson', 'Student ID': 'GK1_KA1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Kevin Martelle', 'Student ID': 'GK1_KM1', 'Student Grade Level': 'K-1'}  ... displaying 10 of 11 total bound parameter sets ...  {'Student Name': 'Olsen Le', 'Student ID': 'GK1_OL1', 'Student Grade Level': 'K-1'}, {'Student Name': 'Terry Long', 'Student ID': 'GK1_TL1', 'Student Grade Level': 'K-1'})]
(Background on this error at: https://sqlalche.me/e/14/f405)

In [ ]:
sql1 = '''select * from student_info;'''
cursor.execute(sql1)
for i in cursor.fetchall():
    print(i)

In [61]:
#Closing the connection
cursor.close()
conn.close()